Add `image_label_mismatch` parameter to `docker_container`

RoGryza opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

Fixes #314 
Add `image_label_mismatch` parameter to `docker_container` to specify the behaviour when the labels provided in the ansible config don't match with the labels from the container's image.

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->

docker_container

felixfontein commented on plugins/modules/docker_container.py:
```diff
@@ -394,6 +394,20 @@
+  image_label_mismatch:
+    description:
+      - How to handle labels inherited from the image that are not set explicitly.
+      - When C(ignore), labels that are present in the image but not specified in ansible will be
```
```suggestion
      - When C(ignore), labels that are present in the image but not specified in I(labels) will be
```

felixfontein commented on plugins/modules/docker_container.py:
```diff
@@ -394,6 +394,20 @@
+    description:
+      - How to handle labels inherited from the image that are not set explicitly.
+      - When C(ignore), labels that are present in the image but not specified in ansible will be
+        ignored. This is useful to avoid having to specify the image labels in ansible while keeping
```
```suggestion
        ignored. This is useful to avoid having to specify the image labels in I(labels) while keeping
```

felixfontein commented on plugins/modules/docker_container.py:
```diff
@@ -394,6 +394,20 @@
+      - When C(ignore), labels that are present in the image but not specified in ansible will be
+        ignored. This is useful to avoid having to specify the image labels in ansible while keeping
+        labels I(comparisons) C(strict).
+        When C(fail), if there are labels present in the image which aren't set from ansible, the
```
```suggestion
      - When C(fail), if there are labels present in the image which are not set from I(labels), the
```

felixfontein commented on plugins/modules/docker_container.py:
```diff
@@ -394,6 +394,20 @@
+        When C(fail), if there are labels present in the image which aren't set from ansible, the
+        module will fail. This prevents introducing unexpected labels from the base image.
+      - "B(Warning:) This option is ignored unless C(labels: strict) or C(*: strict) is specified in
+        the I(comparisons option."
```
```suggestion
        the I(comparisons) option."
```

felixfontein commented on plugins/modules/docker_container.py:
```diff
@@ -2403,6 +2419,19 @@ def has_different_configuration(self, image):
                         # expected_healthcheck comparison in this case.
                         continue
 
+                    if key == 'expected_labels' and self.parameters.image_label_mismatch == 'fail':
```
```suggestion
                    if key == 'expected_labels' and compare == 'strict' and self.parameters.image_label_mismatch == 'fail':
```

felixfontein commented on plugins/modules/docker_container.py:
```diff
@@ -2403,6 +2419,19 @@ def has_different_configuration(self, image):
+                        # base_image_mismatch is fail we want raise an error.
+                        image_labels = self._get_image_labels(image)
+                        would_remove_labels = []
+                        for label in image_labels.keys():
```
Iterating over a map iterates over the map keys automatically:
```suggestion
                        for label in image_labels:
```

felixfontein commented on plugins/modules/docker_container.py:
```diff
@@ -2403,6 +2419,19 @@ def has_different_configuration(self, image):
+                        for label in image_labels.keys():
+                            if label not in self.parameters.labels:
+                                # Format label for error message
+                                would_remove_labels.append(repr(label))
```
```suggestion
                                would_remove_labels.append(label)
```

RoGryza replied:
I used `repr` to quote the labels here, since they can contain spaces which would in my opinion make the error message confusing e.g. `Labels: foo bar, qux` instead of `Labels: 'foo bar', 'qux'`. Do you think I can ignore the quotes or maybe just make it more explicit like `"{}".format(label)`

felixfontein commented on plugins/modules/docker_container.py:
```diff
@@ -2403,6 +2419,19 @@ def has_different_configuration(self, image):
+                                would_remove_labels.append(repr(label))
+                        if would_remove_labels:
+                            msg = "Some labels should be removed but are present in the base image. You can set image_label_mismatch to 'ignore' to ignore this error. Labels: {0}"
+                            self.fail(msg.format(', '.join(would_remove_labels)))
```
```suggestion
                            self.fail(msg.format(', '.join(['"%s"' % label for label in would_remove_labels])))
```

felixfontein commented on plugins/modules/docker_container.py:
```diff
@@ -2641,6 +2670,26 @@ def _get_expected_binds(self, image):
+            return {}
+
+        # Can't use get('Labels', {}) because 'Labels' may be present and be None
+        labels = image[self.parameters.client.image_inspect_source].get('Labels')
+        if labels is None:
+            return {}
+        else:
+            return dict(labels)
```
```suggestion
        return image[self.parameters.client.image_inspect_source].get('Labels') or {}
```

felixfontein commented on plugins/modules/docker_container.py:
```diff
@@ -2641,6 +2670,26 @@ def _get_expected_binds(self, image):
 
+    def _get_expected_labels(self, image):
+        if self.parameters.image_label_mismatch == 'ignore':
+            expected_labels = self._get_image_labels(image)
```
```suggestion
            expected_labels = dict(self._get_image_labels(image))
```

felixfontein commented on plugins/modules/docker_container.py:
```diff
@@ -2641,6 +2670,26 @@ def _get_expected_binds(self, image):
         return result
 
+    def _get_expected_labels(self, image):
+        if self.parameters.image_label_mismatch == 'ignore':
```
If `labels` hasn't been specified, labels must not be compared at all:
```suggestion
        if self.parameters.labels is None:
            return None
        if self.parameters.image_label_mismatch == 'ignore':
```

felixfontein commented on plugins/modules/docker_container.py:
```diff
@@ -2641,6 +2670,26 @@ def _get_expected_binds(self, image):
+            expected_labels = self._get_image_labels(image)
+        else:
+            expected_labels = {}
+        if self.parameters.labels:
+            expected_labels.update(self.parameters.labels)
```
Then it's also no longer necessary to check this one:
```suggestion
        expected_labels.update(self.parameters.labels)
```

felixfontein commented on tests/integration/targets/docker_container/tasks/tests/options.yml:
```diff
@@ -2195,9 +2195,167 @@
+    - image_label_mismatch_6.container.Config.Labels.img_label == "base"
+    - image_label_mismatch_7 is not changed
+    - image_label_mismatch_8 is not changed
+    - image_label_mismatch_9 is failed
```
Please also check the error message:
```suggestion
    - image_label_mismatch_9 is failed
    - image_label_mismatch_9.msg == '...'
```

felixfontein commented on tests/integration/targets/docker_container/tasks/tests/options.yml:
```diff
@@ -2195,9 +2195,167 @@
+    - image_label_mismatch_10 is not changed
+    - image_label_mismatch_11 is changed
+
+###################################################################
+# ipc_mode ########################################################
+###################################################################
```
The original header had one more `#` per line:
```suggestion
####################################################################
## ipc_mode ########################################################
####################################################################
```

felixfontein commented on plugins/modules/docker_container.py:
```diff
@@ -394,6 +394,20 @@
+    type: str
+    choices:
+      - 'ignore'
+      - 'fail'
```
```suggestion
      - 'fail'
    default: ignore
    version_added: 2.6.0
```

felixfontein reviewed (commented):
Thanks for your contribution! I've added some comments. Could you also please add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/collection_development_process.html#creating-a-changelog-fragment)? Thanks :)

RoGryza commented:
Thanks! Addressed all the changes. I added [the change fragment](https://github.com/ansible-collections/community.docker/pull/370/commits/f87e3f8231f06063f4bc056d40edbe8bc3290295#diff-c9e10b3c238f766bfdad843d6ad56d05f4318fe6f4f320b9f81980dfd085841d) but I wasn't sure whether to call it `minor_changes` or `bugfixes`

felixfontein commented on changelogs/fragments/370-add-image-label-mismatch.yml:
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+- "docker_container - added ``image_label_mismatch`` parameter"
```
```suggestion
- "docker_container - added ``image_label_mismatch`` parameter (https://github.com/ansible-collections/community.docker/issues/314, https://github.com/ansible-collections/community.docker/pull/370)."
```

felixfontein commented:
> Thanks! Addressed all the changes. I added [the change fragment](https://github.com/ansible-collections/community.docker/pull/370/commits/f87e3f8231f06063f4bc056d40edbe8bc3290295#diff-c9e10b3c238f766bfdad843d6ad56d05f4318fe6f4f320b9f81980dfd085841d) but I wasn't sure whether to call it `minor_changes` or `bugfixes`

It's kind of both ;-) `minor_changes` should definitely be used, what you could do is also add a `bugfixes` one next to `minor_changes`. But just `minor_changes` is also fine (and maybe the best).

felixfontein commented on plugins/modules/docker_container.py:
```diff
@@ -2403,6 +2422,20 @@ def has_different_configuration(self, image):
                         # expected_healthcheck comparison in this case.
                         continue
 
+                    if key == 'expected_labels' and compare == 'strict' and self.parameters.image_label_mismatch == 'fail':
```
```suggestion
                    if key == 'expected_labels' and compare['comparison'] == 'strict' and self.parameters.image_label_mismatch == 'fail':
```

felixfontein commented on tests/integration/targets/docker_container/tasks/tests/options.yml:
```diff
@@ -2194,6 +2194,165 @@
+    image_label_mismatch: ignore
+    labels:
+      img_label: override
+    state: started
```
Makes the test complete a lot faster.
```suggestion
    state: started
    force_kill: yes
```

felixfontein commented on tests/integration/targets/docker_container/tasks/tests/options.yml:
```diff
@@ -2194,6 +2194,165 @@
+    name: "{{ cname }}"
+    image_label_mismatch: ignore
+    labels: {}
+    state: started
```
```suggestion
    state: started
    force_kill: yes
```

felixfontein commented on tests/integration/targets/docker_container/tasks/tests/options.yml:
```diff
@@ -2194,6 +2194,165 @@
+    image_label_mismatch: fail
+    labels:
+      img_label: override
+    state: started
```
```suggestion
    state: started
    force_kill: yes
```

felixfontein commented on plugins/modules/docker_container.py:
```diff
@@ -2403,6 +2422,20 @@ def has_different_configuration(self, image):
+                                would_remove_labels.append(label)
+                        if would_remove_labels:
+                            msg = ("Some labels should be removed but are present in the base image. You can set image_label_mismatch to 'ignore' to ignore"
+                                   "this error. Labels: {0}")
```
```suggestion
                                   " this error. Labels: {0}")
```

felixfontein commented on tests/integration/targets/docker_container/tasks/tests/options.yml:
```diff
@@ -2194,6 +2194,168 @@
+    - image_label_mismatch_7 is not changed
+    - image_label_mismatch_8 is not changed
+    - image_label_mismatch_9 is failed
+    - image_label_mismatch_9.msg == "Some labels should be removed but are present in the base image. You can set image_label_mismatch to 'ignore' to ignore this error. Labels: \"img_Label\""
```
```suggestion
    - >-
      image_label_mismatch_9.msg == "Some labels should be removed but are present in the base image. You can set image_label_mismatch to 'ignore' to ignore this error. Labels: \"img_Label\""
```

felixfontein commented on tests/integration/targets/docker_container/tasks/tests/options.yml:
```diff
@@ -2194,6 +2194,169 @@
+    - image_label_mismatch_8 is not changed
+    - image_label_mismatch_9 is failed
+    - >-
+      image_label_mismatch_9.msg == "Some labels should be removed but are present in the base image. You can set image_label_mismatch to 'ignore' to ignore this error. Labels: \"img_Label\""
```
```suggestion
      image_label_mismatch_9.msg == ("Some labels should be removed but are present in the base image. You can set image_label_mismatch to 'ignore' to ignore this error. " ~ 'Labels: "img_label"')
```

felixfontein commented:
The only CI node left (PR comments) can be ignored, so merging now :)

felixfontein commented:
@RoGryza thanks for contributing this feature+fix :)
